File: src/main/java/FCP.java
import Lexer.LexerForFCP;
import Lexer.Token;
import Parser.Parser;
import Parser.Printer;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import Parser.Expression;

public class FCP {
    static boolean hadError = false;

    public static void main(String[] args) throws IOException {
        run("fefw");
        /*if (args.length > 1) {
            System.out.println("Too many arguments");
            System.exit(1);
        }
        else if (args.length == 1) {
            runFile(args[0]);
        }
        else {
            runREPL();
        }*/
    }

    private static void runFile(String name) throws IOException {
        byte[] bytes = Files.readAllBytes(Paths.get(name));

        if (hadError) {
            System.exit(1);
        }

        run(new String(bytes, Charset.defaultCharset()));
    }

    static void runREPL() throws IOException {
        InputStreamReader input = new InputStreamReader(System.in);
        BufferedReader reader = new BufferedReader(input);

        for(;;) {
            System.out.println(">");
            String line = reader.readLine();

            if (line == null) {
                break;
            }

            hadError = false;

            run(line);
        }
    }

    static void run(String stringOfCode) {
        LexerForFCP lexer = new LexerForFCP(" 1 + 2 * 3 == 1 + 4 - 2");
        lexer.parseString();
        List<Token> tokens = lexer.listOfTokens;
        System.out.print("tokens: ");

        for (Token i : tokens) {
            System.out.print(i.name);
            System.out.print(" ");
        }
        Parser parser = new Parser(tokens);

        Expression exp  = parser.parse();
        Printer printer = new Printer();
        System.out.println(printer.print(exp));

    }

    public static void error(int string, String message) {
        report(string, "" ,message);
    }

    public static void report(int string, String where, String massage) {
        System.out.println("Error occurred in " + string + "line, in" + where + ": " + massage);

        boolean hadError = true;
    }
}

File: src/main/kotlin/FCP.kt
import com.oracle.truffle.api.Truffle
import com.oracle.truffle.api.frame.MaterializedFrame
import com.oracle.truffle.api.frame.VirtualFrame
import lexer.LexerForFCP
import lexer.Token
import parser.MyRootNode
import parser.Parser
import java.io.File
import java.io.BufferedReader

// read x, y; 1: if x == y goto 7 else 2 2: if x < y goto 5 else 3 3: x := x - y goto 1 5: y := y - x goto 1 7: return x

fun main() {
    //for (i in 1..1000) {
        val startTime = System.nanoTime()
        run("read x, y; 1: x := y - 1 if x == y goto 2 else 3 2: return x 3: x := x + 1 goto 2")
        val totalTime = System.nanoTime() - startTime

        println(totalTime)
    //}
}

fun run(source: String) {
    val lexer = LexerForFCP(source)
    lexer.parseString()

    val tokens: List<Token> = lexer.listOfTokens

    val parser = Parser(tokens)
    val tree = parser.getAst()


    val inputData = mutableListOf<Int>()

    for (i in 1..parser.listOfVariables.size) {
        inputData.add(readLine()?.toInt() ?: TODO("incorrect input"))
    }

    println(tree.execute())
}

fun readFile() {
    val bufferedReader: BufferedReader = File("example.txt").bufferedReader()
    val inputString = bufferedReader.use { it.readText() }
    println(inputString)
}
